pkg/ops/miningManager.go

```diff
@@ -121,26 +166,188 @@ func (mgr *MiningMgr) Start(ctx context.Context) {
+	}
+
+	priceTRB = priceTRB.Mul(priceTRB, big.NewInt(tellorCommon.WEI))
+	priceTRB = priceTRB.Div(priceTRB, big.NewInt(tracker.PSRs[tracker.RequestID_TRB_ETH].Granularity()))
+
+	// Big int can't multiple fractions so need to convert the operation into a division.
+	devider := big.NewInt(0).Div(trb, trb)
+	eth := big.NewInt(0).Div(priceTRB, devider)
+	return eth, nil
+}
+
+func (mgr *MiningMgr) txCost() (*big.Int, error) {
+	slotNum, err := mgr.contractGetter.GetUintVar(nil, rpc.Keccak256([]byte("slotProgress")))
+	if err != nil {
+		return nil, errors.Wrap(err, "getting slotProgress")
+	}
+	// This is the price for the last transaction so increment +1
+	// to get the price for next slot transaction.
+	// Slots numbers should be from 0 to 4 so
+	// use mod of 5 in order to save 5 as slot 0.
+	slotNum.Add(slotNum, big.NewInt(1)).Mod(slotNum, big.NewInt(5))
+	txCostID := tellorCommon.PriceTXs + slotNum.String()
+	cost, err := mgr.database.Get(txCostID)
+	if err != nil {
+		return nil, errors.New("getting the tx eth cost from the db")
+	}
+	// No price record in the db yet.
+	if cost == nil {
+		return big.NewInt(0), nil
+	}
+
+	return big.NewInt(0).SetBytes(cost), nil
+}
+
+// saveTXCost calculates the price for a given slot.
+// Since the transaction doesn't include the slot number it was submited for
+// it gets the slot number as soon as the transaction passes and
+// saves it in the database for profit calculations use.
+// TODO[Krasi] To be more detirministic and simplify this
+// should get the `slotProgress` and `gasUsed` from the `NonceSubmitted` event.
+// At the moment there is a slight chance of a race condition if
+// another transaction has passed between checking the transaction cost and
+// checking the `slotProgress`
+// Tracking issue https://github.com/tellor-io/TellorCore/issues/101
+func (mgr *MiningMgr) saveTXCost(ctx context.Context, tx *types.Transaction) {
+	go func(tx *types.Transaction) {
+		receipt, err := bind.WaitMined(ctx, mgr.ethClient, tx)
+		if err != nil {
+			mgr.log.Error("waiting for transaction completion for  calculating transaction cost err:%v", err)
+		}
+		if receipt.Status != 1 {
+			mgr.log.Error("unsuccessful submitSolution transaction, not saving the tx cost in the db err:%v", err)
+			return
+		}
+
+		gasUsed := big.NewInt(int64(receipt.GasUsed))
+		txCost := gasUsed.Mul(gasUsed, tx.GasPrice())
+		slotNum, err := mgr.contractGetter.GetUintVar(nil, rpc.Keccak256([]byte("slotProgress")))
+		if err != nil {
+			mgr.log.Error("getting slotProgress for calculating transaction cost err:%v", err)
+		}
+
+		mgr.log.Debug("saving transaction cost txHash:%v cost GWEI:%v slot:%v", receipt.TxHash.String(), txCost.Int64()/tellorCommon.GWEI, slotNum.Int64())
+		txCostID := tellorCommon.PriceTXs + slotNum.String()
+		_, err = mgr.database.Put(txCostID, txCost.Bytes())
+		if err != nil {
+			mgr.log.Error("saving transaction cost in the db err:%v", err)
+		}
+	}(tx)
+}
+
+func (mgr *MiningMgr) waitProfitable() error {
```

Comment:
will this exit if a new challenge get's passed down?

Reply:
This will block until a SubmitSolution is profitable, 
Checks the current slot, checks the reward and if profitable it will unblock otherwise it waits 1 sec and checks again. Once it unblocks if the solution it submitted is already submitted I assume the transaction will fail and then it will need to calculate the new solution.

Reply:
Yeah, this will need to change. You want it to unblock if a new challenge comes down

Reply:
Completely refactored the code flow and it is not blocked and I think it is much easier to follow.

Please review only the mining manager changes. The changes regarding the logging are from a different PR and will disappear once that other PR is merged.